### PR TITLE
Changes to awsformat-2.1

### PR DESCRIPTION
**Pull request by MGilhespie**
Changes to code since October 2023 should be merged into the awsformat-2.1 branch.

Changes in processing events, fixing issues with fields and field names.
Removed telemetry "cd" variable.
Removed a nesting level of data in telemetry JSON.
Changes to JSON parsing of commands.

Squashed commits on the awsformat-2.1_changes_to_processing_events_and_acks into a single commit on this awsformat-2.1_changes_to_json branch.

### CODE REVIEW

File: src/iotconnect_discovery.c [hunk 1]
```diff
@@ -56,12 +64,18 @@ static bool split_url(IotclDiscoveryResponse *response) {
 }
 
 IotclDiscoveryResponse *iotcl_discovery_parse_discovery_response(const char *response_data) {
-    cJSON *json_root = cJSON_Parse(response_data);
+	cJSON *json_root = cJSON_Parse(response_data);
```
Comment (wtk-evoirin) [inline src/iotconnect_discovery.c:67]: Whitespace error

File: src/iotconnect_discovery.c [hunk 2]
```diff
@@ -119,46 +133,59 @@ IotclSyncResponse *iotcl_discovery_parse_sync_response(const char *response_data
...
-        tmp_value = cJSON_GetObjectItem(sync_res_json, "rc");
-        if (!tmp_value) {
-            response->rc = -1;
-        }
-        tmp_value = cJSON_GetObjectItem(sync_res_json, "at");
-        if (!tmp_value) {
-            response->at = -1;
-        }
+        cJSON *meta_json = cJSON_GetObjectItemCaseSensitive(sync_res_json, "meta");
+		if (!meta_json) {
```
Comment (wtk-evoirin) [inline src/iotconnect_discovery.c:145]: Indent is a bit garbloofed here...

File: src/iotconnect_discovery.c [hunk 3]
```diff
@@ -119,46 +133,59 @@ IotclSyncResponse *iotcl_discovery_parse_sync_response(const char *response_data
...
+
         if (p) {
+
             response->broker.name = safe_get_string_and_strdup(p, "n");
             response->broker.client_id = safe_get_string_and_strdup(p, "id");
             response->broker.host = safe_get_string_and_strdup(p, "h");
             response->broker.user_name = safe_get_string_and_strdup(p, "un");
+            response->broker.port = get_numeric_value_or_default(p, "p", 8883);
+
+#if 0
```
Comment (wtk-evoirin) [inline src/iotconnect_discovery.c:169]: Is there a particular reason to hide this behind an "if 0"?

Does the old code need to stay visible or be available for reversion later?

In any case, I'd prefer to have a comment explaining that fact and/or use a dsecriptive preproscessor define

(e.g. #if DEPRECATED_PRE_AWS or something)

File: src/iotconnect_discovery.c [hunk 4]
```diff
@@ -119,46 +133,59 @@ IotclSyncResponse *iotcl_discovery_parse_sync_response(const char *response_data
...
                     !response->cpid ||
-                    !response->dtg ||
                     !response->broker.host ||
                     !response->broker.client_id ||
-                    !response->broker.user_name ||
-                    // !response->broker.pass || << password may actually be null or empty
-                    !response->broker.sub_topic ||
-                    !response->broker.pub_topic
-                    ) {
+#if 0
```
Comment (wtk-evoirin) [inline src/iotconnect_discovery.c:182]: See above

+ messy indentation

File: src/iotconnect_event.c [hunk 5]
```diff
@@ -54,7 +53,7 @@ static int to_ack_status(bool success, IotConnectEventType type) {
 
 
 static bool iotc_process_callback(struct IotclEventDataTag *eventData) {
-    if (!eventData) return false;
+	if (!eventData) return false;
```
Comment (wtk-evoirin) [inline src/iotconnect_event.c:60]: Indent :)

File: src/iotconnect_event.c [hunk 6]
```diff
@@ -179,20 +167,25 @@ char *iotcl_clone_download_url(IotclEventData data, size_t index) {
...
     cJSON *ver = cJSON_GetObjectItemCaseSensitive(data->data, "ver");
+    if (!cJSON_IsObject(ver)) {
+        ver = data->data; // AWS and 2.1 we presume...
+    }
     if (cJSON_IsObject(ver)) {
-        cJSON *sw = cJSON_GetObjectItem(ver, "sw");
-        if (is_valid_string(sw)) {
-            return iotcl_strdup(sw->valuestring);
-        }
+		cJSON *sw = cJSON_GetObjectItem(ver, "sw");
```
Comment (wtk-evoirin) [inline src/iotconnect_event.c:180]: Indent :D

File: src/iotconnect_event.c [hunk 7]
```diff
@@ -179,20 +167,25 @@ char *iotcl_clone_download_url(IotclEventData data, size_t index) {
...
+			return iotcl_strdup(sw->valuestring);
+		}
     }
     return NULL;
 }
 
 char *iotcl_clone_hw_version(IotclEventData data) {
     cJSON *ver = cJSON_GetObjectItemCaseSensitive(data->data, "ver");
+    if (!cJSON_IsObject(ver)) {
+        ver = data->data; // AWS and 2.1 we presume...
```
Comment (wtk-evoirin) [inline src/iotconnect_event.c:191]: Comment doesn't add to my understanding of this - what does this mean?

File: src/iotconnect_event.c [hunk 8]
```diff
@@ -276,9 +269,18 @@ char *iotcl_create_ack_string_and_destroy_event(
...
 ) {
+    cJSON *j_ack_id;
+
     if (!data) return NULL;
     // already checked that ack ID is valid in the messages
-    char *ack_id = cJSON_GetObjectItemCaseSensitive(data->data, "ackId")->valuestring;
+
+    if(!data->root) return NULL;
+
+	j_ack_id = cJSON_GetObjectItemCaseSensitive(data->root, "ack");
```
Comment (wtk-evoirin) [inline src/iotconnect_event.c:290]: Indents %)

File: src/iotconnect_event.c [hunk 9]
```diff
@@ -276,9 +269,18 @@ char *iotcl_create_ack_string_and_destroy_event(
...
+
+	if (!j_ack_id) return NULL;
```
Comment (wtk-evoirin) [inline src/iotconnect_event.c:292]: Inconsistent null checks (some in curly brackets, some not)

File: src/iotconnect_telemetry.c [hunk 10]
```diff
@@ -109,27 +110,26 @@ IotclMessageHandle iotcl_telemetry_create(void) {
...
-    if (!cJSON_AddStringToObject(msg->parent, "cd", config->telemetry.cd)) goto cleanup_root;
     if (!cJSON_AddNumberToObject(msg->parent, "mt", 0)) goto cleanup_root;
 
-//    msg->root_value = root;
-//    if (!msg->root_value) goto cleanup;
-
-    //create 1st level d object which is msg->root_vlaue
-    msg->root_value = cJSON_AddObjectToObject(msg->parent, "d");
-    if (!msg->root_value) goto cleanup_root;
+#if 0
```
Comment (wtk-evoirin) [inline src/iotconnect_telemetry.c:118]: See above about #if 0

+ indents!

File: src/iotconnect_telemetry.c [hunk 11]
```diff
@@ -109,27 +110,26 @@ IotclMessageHandle iotcl_telemetry_create(void) {
...
-    //add 2nd level d array
-    msg->telemetry_data_array = cJSON_AddArrayToObject(msg->root_value, "d"); //create 2nd d_array in 1st d object
-    if (!msg->telemetry_data_array) goto cleanup_msg;
+	//create 2nd d_array in 1st d object
+	msg->telemetry_data_array = cJSON_AddArrayToObject(msg->root_value, "d");
+#else
+	//create 2nd d_array in 1st d object
+	msg->root_value = NULL;
+    msg->telemetry_data_array = cJSON_AddArrayToObject(msg->parent, "d");
+#endif
```
Comment (wtk-evoirin) [inline src/iotconnect_telemetry.c:130]: Missing NULL check for telemetry_data_array?

**Review by wtk-evoirin — CHANGES_REQUESTED** [review]
Looks good but;

- Cosmetic issues
- Some unclear things
- One actual issue

Great work :)

**Review by pywtk — APPROVED** [review]
No Notes from me

File: src/iotconnect_telemetry.c [hunk 12]
```diff
@@ -109,27 +110,26 @@ IotclMessageHandle iotcl_telemetry_create(void) {
...
+	msg->root_value = cJSON_AddObjectToObject(msg->parent, "d");
+	if (!msg->root_value) goto cleanup_root;
 
-    //add 2nd level d array
-    msg->telemetry_data_array = cJSON_AddArrayToObject(msg->root_value, "d"); //create 2nd d_array in 1st d object
-    if (!msg->telemetry_data_array) goto cleanup_msg;
+	//create 2nd d_array in 1st d object
+	msg->telemetry_data_array = cJSON_AddArrayToObject(msg->root_value, "d");
+#else
+	//create 2nd d_array in 1st d object
```
Comment (wtk-evoirin) [inline src/iotconnect_telemetry.c:126]: comment here is inaccurate

File: src/iotconnect_discovery.c [hunk 13]
```diff
@@ -103,6 +116,7 @@ void iotcl_discovery_free_discovery_response(IotclDiscoveryResponse *response) {
 
 IotclSyncResponse *iotcl_discovery_parse_sync_response(const char *response_data) {
     cJSON *tmp_value = NULL;
```
Comment (oerg866) [inline src/iotconnect_discovery.c:118]: This variable is now unused and will throw a warning.

File: src/iotconnect_discovery.c [hunk 14]
```diff
@@ -119,46 +133,59 @@ IotclSyncResponse *iotcl_discovery_parse_sync_response(const char *response_data
...
-                    !response->broker.pub_topic
-                    ) {
+#if 0
+					!response->broker.pass ||   // FIXME: password field in discovery response, not sync response.  password may actually be null or empty
+					!response->broker.sub_topic ||
+                    !response->broker.pub_topic ||
+#endif
+					!response->broker.user_name
+            ) {
                 // Assume parsing eror, but it could alo be (unlikely) allocation error
```
Comment (oerg866) [inline src/iotconnect_discovery.c:189]: might as well fix these typos :D

File: src/iotconnect_event.c [hunk 15]
```diff
@@ -91,68 +90,57 @@ bool iotcl_process_event(const char *event) {
...
 
     cleanup:
-
     cJSON_Delete(root);
     return status;
 }
 
 char *iotcl_clone_command(IotclEventData data) {
-    cJSON *command = cJSON_GetObjectItemCaseSensitive(data->data, "command");
+    cJSON *command = cJSON_GetObjectItemCaseSensitive(data->root, "cmd");
```
Comment (oerg866) [inline src/iotconnect_event.c:147]: since we set data to root, is this change actually necessary?

**Comment by wtk-evoirin** [comment]
https://github.com/avnet-iotconnect/iotc-c-lib/blob/a8d6b26cde97099a254a7e254fc9e3bb12862288/src/iotconnect_telemetry.c#L91

Memory leak here!!

File: src/iotconnect_telemetry.c [hunk 16]
```diff
@@ -87,9 +87,6 @@ static cJSON *setup_telemetry_object(IotclMessageHandle message) {
     //add telemetry object(id, tg, d_object) to "d" array of root.
     if (!cJSON_AddItemToArray(message->telemetry_data_array, telemetry_array)) goto cleanup_cto;
 
-    //setup the actual telemetry object to be used in subsequent calls
-    message->current_telemetry_object = cJSON_CreateObject();
-    if (!message->current_telemetry_object) goto cleanup_cto;
     message->current_telemetry_object = d3_object;
```
Comment (wtk-evoirin) [inline src/iotconnect_telemetry.c:90]: Shouldn't this now technically be called d2_object?

**Review by nik-markovic — CHANGES_REQUESTED** [review]
We ought to be taking in my 2.1 c-lib code. Seems like duplication of effort.

**Comment by nik-markovic** [comment]
Cosing in favor of the new 2.1 implementation.

**Pull request closed without merging**